flowzone.yml

```
@@ -1617,12 +1620,20 @@ jobs:
           node-version: 18
 
       - name: Docusaurus Builder
-        uses: product-os/docusaurus-builder@master
+        if: |
+          needs.project_types.outputs.has_readme == 'true' &&
+          inputs.cloudflare_website != '' && inputs.custom_website_build == ''
+        uses: product-os/docusaurus-theme@master
         with:
           repo: ${{ github.event.repository.name }}
           org: ${{ github.repository_owner }}
           default_branch: ${{ github.event.repository.default_branch }}
-          url: https://${{ inputs.cloudflare_website }}.pages.dev
+          url: https://${{ inputs.cloudflare_website }}.pages.dev/
+
+      - name: Custom Website Builder
+        if: |
+          inputs.custom_website_build != ''
+        run: ${{ inputs.custom_website_build }}
```

```suggestion
        run: npm run doc --if-present
```
Why not reuse the same npm script we support here?
https://github.com/product-os/flowzone/blob/740bed380dbaf3c3619bb470e75901a4ae8eb5bc/flowzone.yml#L1029

Then we could avoid a new input and just call an npm script if it exists.

It doesn't even need to be `run doc` but something appropriate we could call if it exists rather than a free form command that is prone to escape and special character issues.

---

I saw that, and `npm run doc` deploys to GitHub pages. Hence, chose to make my own that deploys to CF pages and keep the 2 separate. I will look into the implementation if that's not true.

---

My point is an optional npm script is preferable over a new free-form command input. You can call the script whatever you would like and Flowzone can be configured to run it if it exists.

---

Right on, but aren't we then blocking anyone else who is not using an NPM repo but wants to build their own site? 
The free-form command although rogue serves more people's preferences.

---

Do we have any examples of building sites that do not use npm?

I think a free-form field is more likely to introduce issues with quoting, escaping characters, etc.

---

> Do we have any examples of building sites that do not use npm?

Not at the moment. 

> I think a free-form field is more likely to introduce issues with quoting, escaping characters, etc.

Wouldn't those issues be issues of the users who are implementing it on their repository? Flowzone is providing a custom command. It will run what you provide essentially.

---

We want this project to be opinionated, I think allowing any arbitrary shell command is asking for trouble, when we can assume 99% of use cases would work with an npm script.

---

I agree on that one, I will get this setup as `npm run deploy` or actually go look into `npm run doc` first

---

Do we have a use case right now for a "custom" web site build? If not, perhaps park it until we do and focus on a custom domain (instead of page.dev), which is likely to be the more frequent use case. But maybe I a missing something..

---

Considering what we talked about above @klutchell ^ Should we use this new input `inputs.custom_website_build`  being defined as a boolean instead for Flowzone to know what to do? 

If input present then 
    Custom
if not then 
    default?

---

@ab77 My PR description describes why we need a custom website build. It's to unblock balenaCloud docs deployment to CF pages.

---

Thanks @vipulgupta2048. I also think using npm workflow instead of arbitrary commands is the more correct way to do this and perhaps this workflow can be combined with the existing `npm docs ..`? Though I am not sure what use case that handles. Maybe a different one.

---

After reviewing that, I have started a conversation about this here: https://balena.zulipchat.com/#narrow/stream/348930-product-os.2Fflowzone/topic/Multiple.20project_types

I think we should drop the existing npm run docs functionality and resolve to using CF pages for docs. I can do that in a different PR. For now , 

I am choosing to go with a custom input that decides whether or not we are going custom and if we are then we run `npm run deploy`

---

> I am choosing to go with a custom input that decides whether or not we are going custom and if we are then we run npm run deploy

@vipulgupta2048 `deploy` script as in `npm run deploy` seems ambigious. Deploy what? `docs` seems more clear..

cc @klutchell

---

Do you mean `npm run docs` or `npm run deploy-docs` @ab77 because then the former runs the risk of sounding ambigous? 

Just to say again, this is only for one repository atm. Anyone else using it would be suspicious and needs to prove that they are indeed special or not. Mentioning this to prevent bikeshedding.